src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -537,6 +538,8 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
 
 	/* modify tasks are always assigned using first-replica policy */
 	workerJob->taskList = FirstReplicaAssignTaskList(taskList);
+
+	AssignDistributedTransactionId();
```

Comment:
I think having this at the beginning of the function makes it easier to read / follow. (at least to me)

And also, we should remove `AssignDistributedTransactionId ()` from `BeginCoordinatedTransaction()`. Because with this change we're decoupling distributed transaction ids from coordinated transactions. And, that's actually a big thing.

I'd even implement a new function called `CitusBeginScan()` which simply calls `AssignDistributedTransactionId();`. And, later I'd call that function in both `CitusSelectBeginScan()` and `CitusModifyBeginScan ()` so that we could potentially add more stuff there in the future. Or, if someone implements another `___Scan__()` which could call that function. Anyway, we could skip that for now.

Reply:
> And also, we should remove AssignDistributedTransactionId () from BeginCoordinatedTransaction(). Because with this change we're decoupling distributed transaction ids from coordinated transactions. And, that's actually a big thing.

No, that'd be wrong due to DDLs/TRUNCATE etc. So, not remove from there 😄